File: backend/internal/models/group.go
package models

import (
	"time"
)

// GroupRole represents a user's role within a group
type GroupRole string

const (
	GroupRoleAdmin  GroupRole = "admin"
	GroupRoleMember GroupRole = "member"
)

// Group represents a collection of users (family, friends, team, etc.)
type Group struct {
	ID          string    `gorm:"type:text;primaryKey" json:"id"`
	Name        string    `gorm:"type:text;not null" json:"name"`
	Description *string   `gorm:"type:text" json:"description,omitempty"`
	CreatedBy   string    `gorm:"type:text;not null" json:"created_by"`
	CreatedAt   time.Time `gorm:"type:datetime;not null" json:"created_at"`
	UpdatedAt   time.Time `gorm:"type:datetime;not null" json:"updated_at"`

	// Relationships
	Creator User          `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
	Members []GroupMember `gorm:"foreignKey:GroupID" json:"members,omitempty"`
}

func (Group) TableName() string {
	return "groups"
}

// GroupMember represents a user's membership in a group
type GroupMember struct {
	ID       string     `gorm:"type:text;primaryKey" json:"id"`
	GroupID  string     `gorm:"type:text;not null;uniqueIndex:idx_group_user_active,where:left_at IS NULL" json:"group_id"`
	UserID   string     `gorm:"type:text;not null;uniqueIndex:idx_group_user_active,where:left_at IS NULL" json:"user_id"`
	Role     GroupRole  `gorm:"type:text;not null;default:member" json:"role"`
	JoinedAt time.Time  `gorm:"type:datetime;not null" json:"joined_at"`
	LeftAt   *time.Time `gorm:"type:datetime" json:"left_at,omitempty"`

	// Relationships
	Group Group `gorm:"foreignKey:GroupID" json:"-"`
	User  User  `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

func (GroupMember) TableName() string {
	return "group_members"
}

// IsActive returns true if the member hasn't left the group
func (gm *GroupMember) IsActive() bool {
	return gm.LeftAt == nil
}

// IsAdmin returns true if the member has admin role
func (gm *GroupMember) IsAdmin() bool {
	return gm.Role == GroupRoleAdmin
}

// GroupInvite represents an invitation code to join a group
type GroupInvite struct {
	ID        string    `gorm:"type:text;primaryKey" json:"id"`
	GroupID   string    `gorm:"type:text;not null;index" json:"group_id"`
	Code      string    `gorm:"type:text;not null;uniqueIndex" json:"code"`
	CreatedBy string    `gorm:"type:text;not null" json:"created_by"`
	ExpiresAt time.Time `gorm:"type:datetime;not null" json:"expires_at"`
	MaxUses   int       `gorm:"not null;default:1" json:"max_uses"`
	UseCount  int       `gorm:"not null;default:0" json:"use_count"`
	CreatedAt time.Time `gorm:"type:datetime;not null" json:"created_at"`

	// Relationships
	Group   Group `gorm:"foreignKey:GroupID" json:"group,omitempty"`
	Creator User  `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
}

func (GroupInvite) TableName() string {
	return "group_invites"
}

// IsValid returns true if the invite can still be used
func (gi *GroupInvite) IsValid() bool {
	return time.Now().Before(gi.ExpiresAt) && gi.UseCount < gi.MaxUses
}

// IsExpired returns true if the invite has expired
func (gi *GroupInvite) IsExpired() bool {
	return time.Now().After(gi.ExpiresAt)
}

// HasUsesRemaining returns true if the invite hasn't reached max uses
func (gi *GroupInvite) HasUsesRemaining() bool {
	return gi.UseCount < gi.MaxUses
}

File: ios/Feats/Feats/Models/Group.swift
import Foundation

struct Group: Codable, Identifiable, Equatable {
    let id: String
    let name: String
    let description: String?
    let createdBy: String
    let createdAt: Date
    let updatedAt: Date
    let members: [GroupMember]?

    enum CodingKeys: String, CodingKey {
        case id, name, description, members
        case createdBy = "created_by"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    static func == (lhs: Group, rhs: Group) -> Bool {
        lhs.id == rhs.id
    }
}

struct GroupMember: Codable, Identifiable, Equatable {
    let id: String
    let groupId: String
    let userId: String
    let role: GroupRole
    let joinedAt: Date
    let user: User?

    enum CodingKeys: String, CodingKey {
        case id, role, user
        case groupId = "group_id"
        case userId = "user_id"
        case joinedAt = "joined_at"
    }
}

enum GroupRole: String, Codable {
    case admin
    case member
}

struct GroupInvite: Codable, Identifiable {
    let id: String
    let groupId: String
    let code: String
    let createdBy: String
    let expiresAt: Date
    let maxUses: Int
    let useCount: Int
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, code
        case groupId = "group_id"
        case createdBy = "created_by"
        case expiresAt = "expires_at"
        case maxUses = "max_uses"
        case useCount = "use_count"
        case createdAt = "created_at"
    }

    var isExpired: Bool {
        Date() > expiresAt
    }

    var hasUsesRemaining: Bool {
        maxUses == 0 || useCount < maxUses
    }

    var isValid: Bool {
        !isExpired && hasUsesRemaining
    }

    var usesDescription: String {
        if maxUses == 0 {
            return "Unlimited uses (\(useCount) used)"
        }
        return "\(useCount)/\(maxUses) uses"
    }
}

struct CreateGroupInviteRequest: Codable {
    var maxUses: Int = 1
    var expiresIn: Int = 168 // 7 days in hours

    enum CodingKeys: String, CodingKey {
        case maxUses = "max_uses"
        case expiresIn = "expires_in"
    }
}

struct CreateGroupRequest: Codable {
    let name: String
    let description: String?
}

struct UpdateGroupRequest: Codable {
    let name: String?
    let description: String?
}

struct UpdateMemberRequest: Codable {
    let role: String
}

struct RedeemInviteRequest: Codable {
    let code: String
}
